### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by dropsrobs**
Tests working but any advices would be great.

### CODE REVIEW

File: spec/arguments_lister_spec.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+require_relative '../src/arguments_lister'
+
+# some different approach
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: Try to avoid such comments. 
Its better to make a comment to your own pull request, than making code comments with your considerations.

File: spec/arguments_lister_spec.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    describe '#proper_order' do
+      it 'sorts arguments in descending order' do
+        expect(lister.proper_order lister.args).to eq(%w(tta tpsc att $%ggt))
+      end
+    end
+
+    describe '#non_alpha_check' do
+      it "changes any non_alpha to '_'" do # so the method non_alpha_check returns changed or nil
+                                           # (gsub! method)
+        expect(input.map{|arg| lister.non_alpha_check arg}).to eq([nil, '_ggt', nil, nil])
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: You methods are changing content of given parameter. You are not using the return value, so you suppose to test it this way.
```ruby
it "changes any non_alpha to '_'" do
  expect{ input.map{|arg| lister.non_alpha_check arg} }.to change{ input }.to(%w(tpsc _ggt att tta))
end
```
I don't know if that code will work without some fixes, but it should be tested this way.

File: spec/formatter_spec.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
+require_relative '../src/formatter'
+
+no_args = []
+one_arg = %w(appt#$%)
+many_args = %w(apt tpa %$##grv vpp)
```
Comment (spidzio) [inline spec/formatter_spec.rb]: Use `let` to define these variables.

File: spec/formatter_spec.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+     end
+   end
+
+   context "when last is 'a'" do
+    it 'prefixes with XXX_' do
+      expect(formatter.first_last_check('Bella')).to eq('XXX_Bella')
+    end
+   end
+  end
+end
```
Comment (spidzio) [inline spec/formatter_spec.rb]: Good usage of context and good descriptions for `it` blocks. I really like it.
Although, you designed your module methods to change value of the passed parameter , so you need to use `change` matcher, just as I wrote before.

File: spec/arguments_lister_spec.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+# some different approach
+
+describe ArgumentsLister do
+    let(:input) { %w(tpsc $%ggt att tta) }
+    let(:lister) { ArgumentsLister.new(input) }
+
+    describe '#initialize' do
+      it 'initializes object with parameters' do
+        expect(lister.args).to eq(input)
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: Like I said you can test `initialize`, but you don't have to.
In our case is actually easy and convenient since you have accessor already.
I would also test if arguments have frozen flag - this is actually what your `initialize` do.

File: spec/arguments_lister_spec.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      end
+    end
+
+    describe '#proper_order' do
+      it 'sorts arguments in descending order' do
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: You don't need to test it here. It will be tested in module.

File: spec/arguments_lister_spec.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        expect(lister.proper_order lister.args).to eq(%w(tta tpsc att $%ggt))
+      end
+    end
+
+    describe '#non_alpha_check' do
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: You don't need to test it here. It will be tested in module.

File: spec/arguments_lister_spec.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+      it "changes any non_alpha to '_'" do # so the method non_alpha_check returns changed or nil
+                                           # (gsub! method)
+        expect(input.map{|arg| lister.non_alpha_check arg}).to eq([nil, '_ggt', nil, nil])
+      end
+    end
+
+    describe '#first_last_check' do
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: You don't need to test it here. It will be tested in module.

File: spec/arguments_lister_spec.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+      it "prepends with 'XXX_' any arguments starting or ending with 'a'" do
+        expect(input.map{ |arg| lister.first_last_check arg }).to eq([nil, nil, 'XXX_att', 'XXX_tta'])
+      end
+     end
+
+    describe '#list' do
+      it 'Combines #proper_order, #non_alpha_check, first_last_check' do
+        expect(lister.list).to eq(%w(tpsc _ggt XXX_att XXX_tta))
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: You are not expecting any specific value returning from the `list` method. Your expectation is to print something to the output. You need to test it this way. Look at this page:
https://www.relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/output-matcher

File: spec/arguments_lister_spec.rb (hunk 1 above)
Comment (dropsrobs) [inline spec/arguments_lister_spec.rb]: Ok, thanks!

File: spec/arguments_lister_spec.rb (hunk 7 above)
Comment (dropsrobs) [inline spec/arguments_lister_spec.rb]: Yeah, but so on methods like #non_alpha_check & #first_last_check. They are also tested in module. But wasn't the goal of that task to test them also, as they are being treated as class of instance?

File: spec/arguments_lister_spec.rb (hunk 5 above)
Comment (dropsrobs) [inline spec/arguments_lister_spec.rb]: Right, good point.

File: spec/arguments_lister_spec.rb (hunk 7 above)
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: Still the same methods. You don't need to test twice the same methods. 
Besides module methods are tested by using `FakeClass` instance with module included, so you are actually testing from inside the class.
Class `ArgumentLister` has only one method: 'list'. Other methods are included with the module. So you need to test is - as it is.
If any other questions about testing modules / classes - ask me directly.

**Comment by dropsrobs** [comment]
Suggestions done.

File: spec/arguments_lister_spec.rb [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        it 'changes nothing' do
+          expect { lister.first_last_check arg_z }.to_not change { arg_z }
+        end
+      end
+      context "when there is 'a' at start or end" do
+        it "prepends with 'XXX_" do
+          expect { lister.first_last_check arg_a }.to change { arg_a }.to(expected_result)
+        end
+      end
+    end
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: Description `#non_alpha_check` and `#first_last_check` should be done in module spec. This spec should only have `#initialize` and `#list` descriptions.

File: spec/arguments_lister_spec.rb [hunk 11]
```diff
@@ -0,0 +1,40 @@
+require_relative '../src/arguments_lister'
+
+describe ArgumentsLister do
+    let(:input) { %w(tpsc $%ggt att tta) }
+    let(:lister) { ArgumentsLister.new(input) }
+
+    describe '#initialize' do
+      it 'initializes object with parameters and unfreezes ARGV' do
+        expect(lister.args.frozen?).to eq(false)
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: Nice

File: spec/arguments_lister_spec.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+      context "when there is 'a' at start or end" do
+        it "prepends with 'XXX_" do
+          expect { lister.first_last_check arg_a }.to change { arg_a }.to(expected_result)
+        end
+      end
+    end
+
+    describe '#list' do
+      it 'combines included module methods and outputs result' do
+        expect { lister.list }.to output("tpsc\n_ggt\nXXX_att\nXXX_tta\n").to_stdout
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: Nice

File: spec/arguments_lister_spec.rb [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    let(:input) { %w(tpsc $%ggt att tta) }
+    let(:lister) { ArgumentsLister.new(input) }
+
+    describe '#initialize' do
+      it 'initializes object with parameters and unfreezes ARGV' do
+        expect(lister.args.frozen?).to eq(false)
+      end
+    end
+
+    describe '#non_alpha_check' do
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: This method should be tested the same way as `#first_last_check` - testing if method changed the parameter.

File: spec/arguments_lister_spec.rb (hunk 11 above)
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: Oh, not nice unfortunately.
You are testing frozen flag, but you are giving not frozen params. You should do it this way.
```ruby
describe '#initialize' do
  subject(:lister) { ArgumentsLister.new(input.freeze) }

  it 'initializes object with parameters and unfreezes parameters' do
    expect(lister.args.frozen?).to be false
    lister.args.each do |arg|
      expect(arg.frozen?).to be false
    end
  end
end
```

File: spec/arguments_lister_spec.rb [hunk 14]
```diff
@@ -0,0 +1,40 @@
+require_relative '../src/arguments_lister'
+
+describe ArgumentsLister do
+    let(:input) { %w(tpsc $%ggt att tta) }
+    let(:lister) { ArgumentsLister.new(input) }
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: `lister` is a subject, so you should define it as subject.
```ruby
subject(:lister) { ArgumentsLister.new(input) }
```

File: spec/arguments_lister_spec.rb (hunk 11 above)
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: One more thing. For checking true/false values do not use the `eq()` matcher but `be(false/true)`.
Look at the available matchers:
https://www.relishapp.com/rspec/rspec-expectations/docs/built-in-matchers

File: spec/arguments_lister_spec.rb (hunk 10 above)
Comment (dropsrobs) [inline spec/arguments_lister_spec.rb]: Actually, this is already done in *Formatter* spec, so I'll remove them from *ArgumentsLister*.

**Comment by dropsrobs** [comment]
Suggested solution done.

File: spec/arguments_lister_spec.rb [hunk 15]
```diff
@@ -0,0 +1,19 @@
+require_relative '../src/arguments_lister'
+
+describe ArgumentsLister do
+  subject(:lister) { ArgumentsLister.new(input.freeze) }
+  let(:input) { %w(tpsc $%ggt att tta) }
+
+  describe '#initialize' do
+    it 'initializes object with parameters and unfreezes ARGV' do
```
Comment (spidzio) [inline spec/arguments_lister_spec.rb]: I can give any argument list to the initializer, not necessary ARGV. Please change spec description to 'unfreezes arguments'

File: spec/formatter_spec.rb [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    end
+  end
+
+  describe '#non_alpha_check' do
+    let(:arg_non) { '%$#ttpa' }
+    let(:arg_alpha) { 'ttpa' }
+    let(:expected_result) { '_ttpa' }
+
+    context 'when alphanumeric' do
+      it 'nothing' do
```
Comment (spidzio) [inline spec/formatter_spec.rb]: 'nothing' means nothing. :)
Please make description more readable.

**Comment by spidzio** [comment]
:+1:
